docs/configuration/authentication/file.md

```diff
@@ -18,59 +18,140 @@ file in the configuration file.
-The best algorithm, [Argon2](https://en.wikipedia.org/wiki/Argon2) does though. It won the
-[Password Hashing Competition](https://en.wikipedia.org/wiki/Password_Hashing_Competition) in 2015 and is now
-considered the best hashing function. There is an open [issue](https://github.com/authelia/authelia/issues/577)
-to add support for this hashing function.
+    $ Password hash: $argon2id$v=19$m=65536$3oc26byQuSkQqksq$zM1QiTvVPrMfV6BVLs2t4gM+af5IN7euO0VB6+Q8ZFs
+
+Full CLI Help Documentation:
+
+```
+Hash a password to be used in file-based users database. Default algorithm is argon2id.
+
+Usage:
+  authelia hash-password [password] [flags]
+
+Flags:
+  -h, --help              help for hash-password
+  -i, --iterations int    set the number of hashing iterations (default 1)
+  -k, --key-length int    [argon2id] set the key length param (default 32)
+  -m, --memory int        [argon2id] set the amount of memory param (in KB) (default 65536)
+  -p, --parallelism int   [argon2id] set the parallelism param (default 4)
+  -s, --salt string       set the salt string
+  -l, --salt-length int   set the auto-generated salt length (default 16)
+  -z, --sha512            use sha512 as the algorithm (defaults iterations to 50000, change with -i)
+```
+
+## Password hash function
+
+The supported hash functions are salted Argon2id (default, version 19 only), and salted SHA512 for backwards compatibility.
+This is determined by the prefix `$argon2id$` and `$6$` respectively, as described in this [wiki page](https://en.wikipedia.org/wiki/Crypt_(C)). 
+
+Although SHA512 is supported default hashes are generated with Argon2id. This is because it is
+not the best hash function, while it is a decent algorithm given the number of rounds is big enough the difficulty to 
+crack the hash is not determined by the performance of the machine. The best current algorithm, 
+[Argon2](https://en.wikipedia.org/wiki/Argon2) does though. It won the 
+[Password Hashing Competition](https://en.wikipedia.org/wiki/Password_Hashing_Competition) in 2015 and is currently
+considered the best hashing function. This was implemented due to community feedback as you can see in this closed
+ [issue](https://github.com/authelia/authelia/issues/577).
+ 
+ ### Password hash algorithm tuning
+ 
+ All algorithm tuning is supported for Argon2id. The only configuration variables that affect SHA512
+ are iterations and salt length. The configuration variables are unique to the file authentication provider, thus they all
+ exist in a key under the file authentication configuration key called `password_hashing`. We have set what are considered 
+ as sane and recommended defaults, if you're unsure about which settings to tune, please see the parameters below, or 
+ for a more in depth understanding see the referenced documentation.
+ 
+ #### Argon2 Links
+ [Go Documentation](https://godoc.org/golang.org/x/crypto/argon2)
+ 
+ [IETF Draft](https://tools.ietf.org/id/draft-irtf-cfrg-argon2-05.html)
+ 
```

Comment:
I'd add this link here: https://www.twelve21.io/how-to-choose-the-right-parameters-for-argon2/